keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1301 @@
+
+### User Stories (Optional)
+
+#### Story 1
+
+A cluster administrator upgrades a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first and second nodes are upgraded, any components that were leaders
+  will typically lose the lease during the node downtime
+  - If one happens to retain its lease, it will be preempted by the coordinated
+    election controller after it updates its identity lease with new version
+    information
+- When the third node is upgraded, all components will be at the new version and
+  one will be elected
+
+#### Story 2
+
+A cluster administrator rolls back a cluster's control plane node-by-node,
+expecting version skew to be respected.
+
+- When the first node is rolled back, any components that were leaders will
+  typically loose the lease during the node downtime
+- Once one of the components updates its identity lease with new version
+  information, the coordinated election controller will preempt the current
+  leader so that this lower version component becomes leader.
+- When the remaining two nodes can rollback, the first node will typically
+  remain leader, but if a new election occurs, the available older version
+  components will be elected.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Risk: Amount of writes performed by leader election increases substantially
+
+This enhancement introduces an identity lease for each instance of each
+component.
+
+Example:
+
+- HA cluster with 3 control plane nodes
+- 3 elected components (kube-controller-manager, schedule,
+  cloud-controller-manager) per control plane node
+- 9 identity leases are created and renewed by the components
+
+Introducing this feature is roughtly equivalent to adding the same lease load as
+adding 9 nodes to a kubernetes cluster.
+
+The [API Server Identity enhancement](../1965-kube-apiserver-identity) also
+introduces similar leases. For comparison, in a HA cluster with 3 control plane
+nodes, API Server Identity adds 3 leases.
+
+This risk can be migitated by scale testing and, if needed, extending the lease
+duration and renewal times to reduce writes/s.
+
+#### Risk: lease watches increase apiserver load substantially
+
+The [Unknown Version Interoperability Proxy (UVIP)
+enhancement](../4020-unknown-version-interoperability-proxy) also adds lease
+watches on [API Server Identity](../1965-kube-apiserver-identity) leases in the
+kube-system namespace. This enhancement would increase the expected number of
+resources being watched from ~3 (for UVIP) to ~12.
```

Comment:
I don't understand why 12. Given that every "lease candidate" is now watching it lease, we're effectively talking about having a watch for every "lease candidate". So with CP with 3 instances, it's effectively 3 per component - and there are usually visibly more than 3 components.

Reply:
Updated, yes you're correct this was outdated.